worlds/zork_grand_inquisitor/data_funcs.py

```diff
@@ -0,0 +1,256 @@
+    }
+
+
+def id_to_goals() -> Dict[int, ZorkGrandInquisitorGoals]:
+    return {goal.value: goal for goal in ZorkGrandInquisitorGoals}
+
+
+def id_to_items() -> Dict[int, ZorkGrandInquisitorItems]:
+    return {data.archipelago_id: item for item, data in item_data.items()}
+
+
+def id_to_locations() -> Dict[int, ZorkGrandInquisitorLocations]:
+    return {
+        data.archipelago_id: location
+        for location, data in location_data.items()
+        if data.archipelago_id is not None
+    }
+
+
+def item_groups() -> Dict[str, Set[str]]:
+    groups: Dict[str, Set[str]] = dict()
+
+    tag: ZorkGrandInquisitorTags
+    for tag in ZorkGrandInquisitorTags:
+        groups[tag.value] = set()
+
+    item: ZorkGrandInquisitorItems
+    data: ZorkGrandInquisitorItemData
+    for item, data in item_data.items():
+        if data.tags is not None:
+            for tag in data.tags:
+                groups[tag.value].add(item.value)
+
+    return {k: v for k, v in groups.items() if len(v)}
+
+
+def items_with_tag(tag: ZorkGrandInquisitorTags) -> Set[ZorkGrandInquisitorItems]:
+    items: Set[ZorkGrandInquisitorItems] = set()
+
+    item: ZorkGrandInquisitorItems
+    data: ZorkGrandInquisitorItemData
+    for item, data in item_data.items():
+        if data.tags is not None and tag in data.tags:
+            items.add(item)
+
+    return items
+
+
+def game_id_to_items() -> Dict[int, ZorkGrandInquisitorItems]:
+    mapping: Dict[int, ZorkGrandInquisitorItems] = dict()
+
+    item: ZorkGrandInquisitorItems
+    data: ZorkGrandInquisitorItemData
+    for item, data in item_data.items():
+        if data.statemap_keys is not None:
+            for key in data.statemap_keys:
+                mapping[key] = item
+
+    return mapping
+
+
+def location_groups() -> Dict[str, Set[str]]:
+    groups: Dict[str, Set[str]] = dict()
+
+    tag: ZorkGrandInquisitorTags
+    for tag in ZorkGrandInquisitorTags:
+        groups[tag.value] = set()
+
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+    for location, data in location_data.items():
+        if data.tags is not None:
+            for tag in data.tags:
+                groups[tag.value].add(location.value)
+
+    return {k: v for k, v in groups.items() if len(v)}
+
+
+def locations_by_region(include_deathsanity: bool = False) -> Dict[
+    ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]
+]:
+    mapping: Dict[ZorkGrandInquisitorRegions, Set[ZorkGrandInquisitorLocations]] = dict()
+
+    region: ZorkGrandInquisitorRegions
+    for region in ZorkGrandInquisitorRegions:
+        mapping[region] = set()
+
+    location: ZorkGrandInquisitorLocations
+    data: ZorkGrandInquisitorLocationData
+    for location, data in location_data.items():
+        if include_deathsanity is False and ZorkGrandInquisitorTags.DEATHSANITY in (
+            data.tags or tuple()
+        ):
+            continue
+
+        mapping[data.region].add(location)
```

Comment:
```suggestion
        mapping.setdefault(data.region, set()).add(location)
```

Reply:
This is not equivalent as the other place where you suggested to use `setdefault`. Menu region doesn't have locations, so it breaks. Keeping this the way it is.